### PR TITLE
RFC: Make inbounds macros expression-like more often

### PR DESCRIPTION
**Pull request by mbauman**
With #11169 fixed, we can now make `@inbounds` return its value ~~*edit: in common cases*.~~

### CODE REVIEW

**Comment by KristofferC** [comment]
Seems like it could be worth to run the benchmarks? If Nanosoldier is feeling better now that is.

**Comment by eschnett** [comment]
I would wait for #15548 to ensure that this behaviour doesn't go away.

**Comment by mbauman** [comment]
> ensure that this behaviour doesn't go away.

That's part of the plan.  If this gets merged, I don't think there's any going back.  :)

Sure, let's do benchmarks… touching these basic utilities can sometimes do surprising things.  What are the categories these days?  Is SIMD its own category?

**Comment by mbauman** [comment]
@nanosoldier `runbenchmarks("array")`

**Comment by jrevels** [comment]
The error on nanosoldier is

```
ERROR: LoadError: UndefVarError: A_ldiv_B not defined
```

@andreasnoack  I believe the [relevant benchmark](https://github.com/JuliaCI/BaseBenchmarks.jl/blob/master/src/linalg/LinAlgBenchmarks.jl#L39) is just testing the `\` operator...did something change there recently? I searched all of Base on [the merge commit the benchmarks ran on](https://github.com/JuliaLang/julia/commit/b4bfebef714af47c6e039bdb240804bbab632013) for a typo (i.e. `A_ldiv_B` instead of `A_ldiv_B!`) but didn't see anything.

**Comment by andreasnoack** [comment]
I think the error is [here](https://github.com/JuliaLang/julia/blob/master/base/linalg/triangular.jl#L1348). I'm surprised that we don't have coverage for those methods. I'll take a look.

**Comment by timholy** [comment]
What needs to be done here?

**Comment by mbauman** [comment]
Is nanosoldier feeling any better these days?  @nanosoldier `runbenchmarks("array")`

Or we can just press merge.  I don't foresee this causing trouble.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-224427554) has completed - successfully executed benchmarks. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/29ab35fbdb9b4d676e2216893dea20d41a9417f4/0fb7006/0fb7006.md). cc @jrevels

**Comment by yuyichao** [comment]
Maybe you missed the `vs` part?

**Comment by mbauman** [comment]
Doh.  I saw it was optional but I figured the default was against master.  Let's try that again: @nanosoldier `runbenchmarks("array", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-224441399) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/3603df487d891aeb106bf6701e856b74c4ab04fb/d8938a5/d8938a5_vs_d98f2c0.md). cc @jrevels

**Comment by timholy** [comment]
Surprising number of performance regressions. Might be fixed by a simple rebase?

**Comment by mbauman** [comment]
Marking this as request-for-help.  This change is very simple and would be very nice to merge, but I'm not well-equipped to debug this sort of regression.

**Comment by mbauman** [comment]
Aha, I now see https://github.com/JuliaLang/julia/issues/16894 — I bet that's what's going on here.

**Comment by mbauman** [comment]
@nanosoldier `runbenchmarks("array", vs = ":master")`?

**Comment by mbauman** [comment]
@nanosoldier `runbenchmarks("array", vs = ":master")`

**Comment by tkelman** [comment]
I think nanosoldier is still being updated to work properly in a world where master is 0.6-dev. Too bad we lost track of this one, should've given it another go after #16897 was merged.

**Comment by jrevels** [comment]
Alright, our dear Nanosoldier should be back on its feet:

@nanosoldier `runbenchmarks("array", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-238420692) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6b91eea256793db9cca0628537ff11ddabceb90e/3652774_vs_2037464/report.md). cc @jrevels

**Comment by vchuravy** [comment]
Should we re-run nanosoldier one last time and merge this, if it comes back OK?

@nanosoldier `runbenchmarks("array", vs = ":master")`

**Comment by jrevels** [comment]
[Nanosoldier stopped working through it's job queue for some reason while I was away](https://github.com/JuliaLang/julia/pull/18793#issuecomment-252758424). Restarted the server:

@nanosoldier `runbenchmarks("array", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-252759105) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2ebe973b59367306b52b37865ed0c23967f49ced/1d384a9_vs_2b7481d/report.md). cc @jrevels

**Comment by mbauman** [comment]
Hm, those regressions look real. I would bet that this is pushing the inlining heuristic over the edge in some cases.

**Comment by vchuravy** [comment]
Interestingly the daily benchmark showed a bunch of improvements in the bitarray benchmark https://github.com/JuliaCI/BaseBenchmarkReports/blob/4c71955ce4d30d4938b34d5a74abfa64caf74a89/daily_2016_10_12/report.md

I am wondering if it was inline heuristics, then this might have helped.
@nanosoldier `runbenchmarks("array", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-253127268) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/5d454c11f10ab8313f724fc71aa7258af0cdf924/82d6438_vs_11d3c96/report.md). cc @jrevels

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks("array", vs = ":master")`

**Comment by jrevels** [comment]
Nanosoldier feel over last night due to a network error, just restarted it.

@nanosoldier `runbenchmarks("array", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-265461067) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/cee112fda61371a1a5a9ff4dc0dc6ab701da1776/9e1151c_vs_934ad48/report.md). cc @jrevels

**Comment by KristofferC** [comment]
Still not completely resolved it seems.

**Comment by mbauman** [comment]
Alright, I found the cause of the regressions.  Code with `@inbounds begin` or `@inbounds if` blocks (like [glue_src_bitchunks](https://github.com/JuliaLang/julia/blob/07cb7c19a5691df36a9d45ea486343e126fd8657/base/bitarray.jl#L87-L93)) return a value that's type-unstable.  This now assigns that unstable value to an unused variable.  Adding an explicit `nothing` value at the end of the block fixes the BitArray regressions.

Can we make inference more robust to unused type-unstable variables?

**Comment by blakejohnson** [comment]
@mbauman good find. Do you think the `cumsum!` regressions are related to the same problem?

**Comment by mbauman** [comment]
I could not reproduce those locally.

**Review by StefanKarpinski — DISMISSED** [review]
LGTM, as long as it doesn't tank performance. It's certainly more convenient to use like this.

**Comment by mbauman** [comment]
Bump.  Any chance this can make it into 0.6?  I suppose we could go through all `@inbounds begin` and `@inbounds if` blocks to ensure that they evaluate to a type stable value… but it'd really be nice to address the inference issue directly.

**Comment by tkelman** [comment]
d'oh, we forgot to cover this in the triage call.

> Adding an explicit `nothing` value at the end of the block fixes the BitArray regressions.

Where does that need to be done?

**Comment by mbauman** [comment]
That particular spot is after the `if` block here: https://github.com/JuliaLang/julia/blob/07cb7c19a5691df36a9d45ea486343e126fd8657/base/bitarray.jl#L92. But there are surely other places that aren't covered by perf tests. There are quite a few `@inbounds (begin|if)` blocks in base, and probably even more ternary expressions. It's a slog.

**Comment by mbauman** [comment]
Tentatively tagging this as 0.6 so it doesn't get lost again.  It'd be nice if @vtjnash or @JeffBezanson could comment on the feasibility of improving codegen in the case of an unused type-unstable variable.  If that is hard or unlikely to happen, then I can audit the code in Base to limit the impact, but it'd still affect packages… and by definition it'd affect them in performance sensitive code.

Minimal repro:

```jl
julia> function f(b)
           y = 1
           x = if b
               y = 2
           end
           x
           return y
       end

julia> @code_llvm f(true)

define i64 @julia_f_61844(i8) #0 {
top:
  %1 = call %jl_value_t*** @jl_get_ptls_states() #2
  …
```

**Comment by StefanKarpinski** [comment]
@JeffBezanson, @vtjnash – can you please take a look at this?

**Comment by JeffBezanson** [comment]
The reduced test case is very helpful, but I fear not entirely accurate. Closer would be

```jl
julia> function f(b)
           y = 1
           x = if b
               y = 2
           end
           x
           return y
       end
```

In this macro, the variable is actually used in some sense, so proving it unused is harder.

**Comment by tkelman** [comment]
I think we'll have to defer this then, if compiler optimizations to make it acceptable aren't immediately forthcoming.

**Comment by mbauman** [comment]
How about this compromise?  `@inbounds` returns its value in simple cases.  `@inbounds begin`, `@inbounds if` and `@inbounds (x; y; z)` will still always return nothing.  That'll prevent the bad performance issues, and it'll make `@inbounds` work as most folks want in the most common cases.

**Comment by blakejohnson** [comment]
@mbauman I like that as a stopgap measure.

File: base/essentials.jl
Comment (tkelman) [inline base/essentials.jl]: 6 spaces is an unconventional choice :)

Comment (mbauman) [inline base/essentials.jl]: Hah, looks like my editor was thrown off by the wrapped block spacing.  Good catch, thanks.

**Comment by blakejohnson** [comment]
LGTM

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks("array")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-270747215) has completed - successfully executed benchmarks. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/87ea5cd11567ded8936faa196c749bfce653a97d/86b8e42/report.md). cc @jrevels

**Comment by KristofferC** [comment]
Oh.. Woops.

**Comment by mbauman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-270771427) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/313fdb42176ecde5470fbe838cbdf4892a5df58b/86b8e42_vs_d7625dd/report.md). cc @jrevels

**Comment by mbauman** [comment]
Still not perfect.  The increase in allocations in the last remaining perf test is from this line: [`@inbounds (x, i) = next(itr, i)`](https://github.com/JuliaLang/julia/blob/5b6ff4afdcc6c828570be0b8a8d5637081a1b4e4/base/reduce.jl#L45).  That line gets the expression-like treatment… and while it's type stable, it ends up assigning a tuple to an unused variable… which ends up allocating instead of just immediately getting destructured.  Bah.

**Comment by Keno** [comment]
The reduced examples above are efficient now, so it might be worth trying this again in the fully general way.

**Comment by yuyichao** [comment]
Not sure if any important cases still have performance regression but https://github.com/JuliaLang/julia/pull/23240 should make all dummy assignments and unused variables free.

**Comment by mbauman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-325245067) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8c1434144844ca4896995ccc7dffa043a8589c40/758fe62_vs_9ac2b57/report.md). cc @ararslan

**Comment by mbauman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-325677284) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/638371928138e0603f51bb4dea0c667000c315df/6a6738a_vs_6a47db8/report.md). cc @ararslan

**Comment by mbauman** [comment]
Ok, well, I've identified a place where I'm still seeing an extra allocation with the expression-like behavior.  See https://github.com/JuliaLang/julia/pull/15558/commits/0f2b7fdaf5d15496d9f1004242571db2dde3b8b7 for the workaround.  I'm having a hard time reducing it down.

**Comment by mbauman** [comment]
Alright, let's see where we are again here.  I may still need to do some work on the inlining cost calculation.

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-346449331):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8ab7530959b0a07a3fa7445339ca1c8c90b4e810/25f938b_vs_80d8719/report.md)
cc @ararslan

**Comment by ararslan** [comment]
Looks like there was some internal server error. I restarted the server, so let's try again: @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by ararslan** [comment]
Yeah okay something is wrong with Nanosoldier.

**Comment by ararslan** [comment]
I think Nanosoldier is feeling better now, so @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-346921538):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/27733d01d24f23c5a8f7650ab8e102228f323d03/b38bfef_vs_f8243b7/report.md)
cc @ararslan

**Comment by ararslan** [comment]
Sorry, still a few more things to fix in Nanosoldier.

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-346991537):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/bd09b945eb975a0c751ec80802fb5532286c98b6/6faaf62_vs_e64ad77/report.md)
cc @ararslan

**Comment by ararslan** [comment]
Sorry for the noise here. Dates moving to the stdlib broke basically the entire Nanosoldier stack and there have still been a couple of things lurking. I think I've _finally_ gotten them all, but we'll see.

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15558#issuecomment-347033386) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8bb9cbd7a83105a0cb5405bbb6dfce7ce620b86c/e4355c5_vs_061593e/report.md). cc @ararslan

**Comment by blakejohnson** [comment]
Looks like a few real memory regressions?

**Comment by mbauman** [comment]
Yes, thanks for fixing up Nanosoldier for this, Alex!

It looks like we're still at the same place that stymied me last time — I worked around it before by changing https://github.com/JuliaLang/julia/blob/690f4bfdafbc9cda2ccc7c16eec01cc73ac108a4/base/subarray.jl#L162 from:

        @inbounds idxs = to_indices(V.parent, reindex(V, V.indexes, I))

to:

    idxs = @inbounds to_indices(V.parent, reindex(V, V.indexes, I))

It looks like the use of an intermediary variable here causes a tuple to be allocated.

**Comment by StefanKarpinski** [comment]
Let's go ahead and merge this with the workaround now and then once https://github.com/JuliaLang/julia/pull/24113 is merged, verify that the workaround is no longer needed.

**Comment by StefanKarpinski** [comment]
This seems to be good to go and passes CI. Shall we go ahead and merge this?

**Comment by JeffBezanson** [comment]
Already merged as #25033

**Pull request closed without merging**
